### PR TITLE
Include CPU in darwin platform string

### PR DESCRIPTION
**Pull request by keith**
Bazel uses the CPU to determine the --apple_platform_type to build for
in some cases. Previously on a macOS machine the CPU would be returned
as just `darwin`, which isn't considered a valid macOS CPU, instead we
should return what we support and include the architecture.

### CODE REVIEW

**Comment by keith** [comment]
cc @philwo since you were mentioned on the TODO

**Comment by keith** [comment]
An alternative (or addition to this) could be to add `darwin` the the list of macOS CPUs here https://github.com/bazelbuild/bazel/blob/0ccee07ffca4c2da0788efdd01c9a724d3e94cc1/src/main/java/com/google/devtools/build/lib/rules/apple/ApplePlatform.java#L55-L56 let me know if that would make sense to do too

**Comment by keith** [comment]
It looks like some other stuff also assumes it's just `darwin` https://github.com/bazelbuild/bazel/blob/edf737592d5e75f43c93f51cf66bb3b15e88d7c5/tools/cpp/lib_cc_configure.bzl#L180-L181 which also appears to be causing the test failures. I'm leaning more towards adding `darwin` to `MACOS_TARGET_CPUS` but I'm still debugging the tests

**Comment by keith** [comment]
I pushed a possible fix here, I'm not entirely sure this is the right option. Running this locally doesn't actually work but I believe it's because the tests invoke my globally installed bazel somehow, instead of the one that was just build, and therefore get the old `local_config_cc`

**Comment by keith** [comment]
Which one should I unify on though? Likely unifying on the current one would be easier. But either way removing one will be a breaking change I guess? Specifically this might break people who have config that selects on Darwin as the cpu? I was still debugging why the tests are failing, but now I'll wait for a decision here before fixing them. 


--
Keith Smiley

**Comment by keith** [comment]
@trybka while working on this fix I'm hitting a new issue that @sergiocampama thinks you might be able to help with.

This test:

https://github.com/bazelbuild/bazel/blob/682a0d186f4a47e74f8aaa8132b84f51c1778a4f/src/test/java/com/google/devtools/build/lib/rules/cpp/CcCommonTest.java#L690-L702

Fails on macOS with this change because of this conditional:

https://github.com/bazelbuild/bazel/blob/00d2050c88c19bed85bb34127b1147e1486b67c9/src/main/java/com/google/devtools/build/lib/rules/cpp/CcCommon.java#L249-L252

Before my change `ApplePlatform.isApplePlatform("darwin")` would return false, which seemed to have incorrectly made this test pass. Now that I'm fixing that by changing the CPU in this case to one that `isApplePlatform` returns `true` for, this test fails.

So I have 2 questions:

1. Why can't apple platforms support static `cc_library` targets?
2. If there's a un-fixable reason for 1, how should I fix this test?

**Comment by keith** [comment]
@aragos I've updated this PR to only support `darwin_x86_64` in case we have to support other architectures in the future. We'll see if there are any new test failures besides the ones mentioned above

File: tools/cpp/crosstool_lib.bzl [hunk 1]
```diff
@@ -284,7 +284,7 @@ def _is_linux(platform):
     return platform == "k8"
 
 def _is_darwin(platform):
-    return platform == "darwin"
+    return platform.startswith("darwin")
```
Comment (keith) [inline tools/cpp/crosstool_lib.bzl]: I might be able to change this to just `darwin_x86_64`, I'm not entirely sure, but also if we eventually have to support a new `darwin`, I assume this is what we'll want

**Comment by keith** [comment]
It looks like the only failing test is the one I commented about above, so once we have a solution to that this should be green

**Comment by keith** [comment]
I've pushed one possible solution for the test failure I asked about above https://github.com/bazelbuild/bazel/commit/7c168440b4184b25647d96ba44e424ccf167f869

This commit separates the failing test for darwin and validates that it errors with that message in this case, instead of having just the warning

**Comment by hlopko** [comment]
I don't think we can accept this without going through incompatible flag. This can (and will) break existing code. Since we are planning to migrate C++ (and objc transitively) rules to platforms (https://www.bazel.build/roadmaps/cpp.html#c-rules-use-platforms-2019q1) I think the better course of action is to make https://source.bazel.build/bazel/+/5609d14ece35fc57084bb66161c7b579cd836c5e:tools/platforms/BUILD more robust, and force users to migrate only once. Wdyt?

**Comment by keith** [comment]
What do you mean by making that platform configuration more robust?

**Comment by keith** [comment]
Also I guess a non-breaking change we could make here instead would be to add `darwin` to the acceptable list of macOS apple platforms. This would fix the underlying issue and just not set us up for the future as much

**Comment by keith** [comment]
I've submitted the `darwin` addition option here: https://github.com/bazelbuild/bazel/pull/7151

**Comment by keith** [comment]
Sounds good, I'm hoping we can merge the alternative fix here https://github.com/bazelbuild/bazel/pull/7151 and then punt on this portion of it until it's a better time

**Comment by keith** [comment]
The other PR has landed, so the original core issue is fixed here. We can revisit this in the future

**Pull request closed without merging**